src/os/bluestore/KernelDevice.cc

```diff
@@ -592,9 +592,17 @@ int KernelDevice::write(
 	   << dendl;
   assert(is_valid_io(off, len));
 
-  if ((!buffered || bl.get_num_buffers() >= IOV_MAX) &&
-      bl.rebuild_aligned_size_and_memory(block_size, block_size)) {
-    dout(20) << __func__ << " rebuilding buffer to be aligned" << dendl;
+  if ((!buffered || bl.get_num_buffers() >= IOV_MAX)) {
+    uint64_t aligned_size = block_size;
+    // Avoid after rebuild_aligned_size_and_memory, get_num_buffers() > IOV_MAX
+    if (bl.get_num_buffers() >= IOV_MAX && (len > (IOV_MAX * block_size))) {
+      aligned_size = ROUND_UP_TO(ROUND_UP_TO(len, IOV_MAX) / IOV_MAX, block_size);
+      dout(5) << __func__ << " because too many buffers, using  " << aligned_size
```

Comment:
i am not quite sure about `dout(5)`. is a large buffer size an unusual thing and we should watch out? actually what i suggested in last comment is to print the `aligned_size` at the "rebuilding buffer to be aligned" line with `dout(20)`.